rust/flexbuffers/src/reader/map.rs

```diff
@@ -78,60 +82,71 @@ impl<'de> MapReader<'de> {
         }
         None
     }
+
     /// Index into a map with a key or usize.
-    pub fn index<I: MapReaderIndexer>(&self, i: I) -> Result<Reader<'de>, Error> {
+    pub fn index<I: MapReaderIndexer>(&self, i: I) -> Result<Reader<B>, Error> {
         i.index_map_reader(self)
     }
+
     /// Index into a map with a key or usize. If any errors occur a Null reader is returned.
-    pub fn idx<I: MapReaderIndexer>(&self, i: I) -> Reader<'de> {
+    pub fn idx<I: MapReaderIndexer>(&self, i: I) -> Reader<B> {
         i.index_map_reader(self).unwrap_or_default()
     }
-    fn usize_index(&self, i: usize) -> Result<Reader<'de>, Error> {
+
+    fn usize_index(&self, i: usize) -> Result<Reader<B>, Error> {
         if i >= self.length {
             return Err(Error::IndexOutOfBounds);
         }
         let data_address = self.values_address + self.values_width.n_bytes() * i;
         let type_address = self.values_address + self.values_width.n_bytes() * self.length + i;
         let (fxb_type, width) = self
             .buffer
+            .as_ref()
             .get(type_address)
             .ok_or(Error::FlexbufferOutOfBounds)
             .and_then(|&b| unpack_type(b))?;
         Reader::new(
-            &self.buffer,
+            self.buffer.clone(),
             data_address,
             fxb_type,
             width,
             self.values_width,
         )
     }
-    fn key_index(&self, k: &str) -> Result<Reader<'de>, Error> {
+
+    fn key_index(&self, k: &str) -> Result<Reader<B>, Error> {
         let i = self.index_key(k).ok_or(Error::KeyNotFound)?;
         self.usize_index(i)
     }
+
     /// Iterate over the values of the map.
-    pub fn iter_values(&self) -> ReaderIterator<'de> {
+    pub fn iter_values(&self) -> ReaderIterator<B> {
         ReaderIterator::new(VectorReader {
             reader: Reader {
-                buffer: self.buffer,
+                buffer: self.buffer.clone(),
                 fxb_type: crate::FlexBufferType::Map,
                 width: self.values_width,
                 address: self.values_address,
             },
             length: self.length,
         })
     }
+
     /// Iterate over the keys of the map.
     pub fn iter_keys(
         &self,
-    ) -> impl Iterator<Item = &'de str> + DoubleEndedIterator + ExactSizeIterator + FusedIterator
+    ) -> impl Iterator<Item = &str> + DoubleEndedIterator + ExactSizeIterator + FusedIterator
     {
-        self.keys_vector().iter().map(|k| k.as_str())
+        unsafe {
+            // TODO(colindjk) is unsafe avoidable here?
+            self.keys_vector().iter().map(|k| (k.as_str() as *const str).as_ref().unwrap())
```

Comment:
I'm not comfortable with this use of unsafe.

I think Rust is complaining that `k.as_str()` returns a string that will not outlive the transient `k`. It doesn't know that that `B` is actually `Bytes` and will keep the reference valid... I think this use of unsafe makes the `InnerBuffer` trait unsafe and I don't want that.